halo2_middleware/src/zal.rs

```diff
@@ -0,0 +1,309 @@
+        ) -> C::Curve;
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve;
+        // Execute MSM according to descriptors
+        // Unsure of naming, msm_with_cached_inputs, msm_apply, msm_cached, msm_with_descriptors, ...
+    }
+}
+
+// ZAL using Halo2curves as a backend
+// ---------------------------------------------------
+
+pub mod impls {
+    use std::marker::PhantomData;
+
+    use crate::zal::traits::MsmAccel;
+    use halo2curves::msm::best_multiexp;
+    use halo2curves::CurveAffine;
+
+    // Halo2curve Backend
+    // ---------------------------------------------------
+    #[derive(Default)]
+    pub struct H2cEngine;
+
+    pub struct H2cMsmCoeffsDesc<'c, C: CurveAffine> {
+        raw: &'c [C::Scalar],
+    }
+
+    pub struct H2cMsmBaseDesc<'b, C: CurveAffine> {
+        raw: &'b [C],
+    }
+
+    impl H2cEngine {
+        pub fn new() -> Self {
+            Self {}
+        }
+    }
+
+    impl<C: CurveAffine> MsmAccel<C> for H2cEngine {
+        fn msm(&self, coeffs: &[C::Scalar], bases: &[C]) -> C::Curve {
+            best_multiexp(coeffs, bases)
+        }
+
+        // Caching API
+        // -------------------------------------------------
+
+        type CoeffsDescriptor<'c> = H2cMsmCoeffsDesc<'c, C>;
+        type BaseDescriptor<'b> = H2cMsmBaseDesc<'b, C>;
+
+        fn get_coeffs_descriptor<'c>(&self, coeffs: &'c [C::Scalar]) -> Self::CoeffsDescriptor<'c> {
+            // Do expensive device/library specific preprocessing here
+            Self::CoeffsDescriptor { raw: coeffs }
+        }
+        fn get_base_descriptor<'b>(&self, base: &'b [C]) -> Self::BaseDescriptor<'b> {
+            Self::BaseDescriptor { raw: base }
+        }
+
+        fn msm_with_cached_scalars(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &[C],
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base)
+        }
+
+        fn msm_with_cached_base(
+            &self,
+            coeffs: &[C::Scalar],
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs, base.raw)
+        }
+
+        fn msm_with_cached_inputs(
+            &self,
+            coeffs: &Self::CoeffsDescriptor<'_>,
+            base: &Self::BaseDescriptor<'_>,
+        ) -> C::Curve {
+            best_multiexp(coeffs.raw, base.raw)
+        }
+    }
+
+    // Backend-agnostic engine objects
+    // ---------------------------------------------------
+    #[derive(Debug)]
+    pub struct PlonkEngine<C: CurveAffine, MsmEngine: MsmAccel<C>> {
+        pub msm_backend: MsmEngine,
+        _marker: PhantomData<C>, // compiler complains about unused C otherwise
+    }
+
+    #[derive(Default)]
+    pub struct PlonkEngineConfig<C, M> {
+        curve: PhantomData<C>,
+        msm_backend: Option<M>,
```

Comment:
Is this ever `None`?  I wonder if this could be `msm_backend: M`.
When it's `NoMsmEngine` it would work because it's an empty struct, and when it becomes `HasMsmEngine` we already have the engine object around; so maybe there's no need to be `Option`?

Reply:
addressed in https://github.com/privacy-scaling-explorations/halo2/pull/277/commits/02c426414de5f9a4f9840a6a9dbfe5f572858813